server/src/main/java/org/opensearch/index/shard/RefreshListeners.java

```diff
@@ -182,23 +201,78 @@ public boolean addOrNotify(Translog.Location location, Consumer<Boolean> listene
+     * @param listener for the refresh. Called with true if registering the listener ran it out of slots and forced a refresh. Called with
+     *        false otherwise.
+     * @return did we call the listener (true) or register the listener to call later (false)?
+     */
+    public void addOrNotifySeqNoRefresh(Long maxSeqNo, Consumer<Boolean> listener) {
+        requireNonNull(listener, "listener cannot be null");
+
+        if (lastMaxSeqNo != SequenceNumbers.NO_OPS_PERFORMED && lastMaxSeqNo >= maxSeqNo) {
+            listener.accept(false);
+            return;
+        }
+        synchronized (this) {
+            if (closed) {
+                throw new IllegalStateException("can't wait for refresh on a closed index");
+            }
+            List<Tuple<Long, Consumer<Boolean>>> listeners = seqNoRefreshListeners;
+            final int maxRefreshes = getMaxRefreshListeners.getAsInt();
+            if (refreshForcers == 0 && maxRefreshes > 0 && (listeners == null || listeners.size() < maxRefreshes)) {
+                Consumer<Boolean> contextPreservingListener = getContextListener(listener);
+                if (listeners == null) {
+                    listeners = new ArrayList<>();
+                }
+                // We have a free slot so register the listener
+                listeners.add(new Tuple<>(maxSeqNo, contextPreservingListener));
+                seqNoRefreshListeners = listeners;
+                return;
+            }
+        }
+        // No free slot so force a refresh and call the listener in this thread
+        listener.accept(true);
+        return;
+    }
+
+    private Consumer<Boolean> getContextListener(Consumer<Boolean> listener) {
+        ThreadContext.StoredContext storedContext = threadContext.newStoredContext(true);
+        return forced -> {
+            try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+                storedContext.restore();
+                listener.accept(forced);
+            }
+        };
+    }
+
     @Override
     public void close() throws IOException {
         List<Tuple<Translog.Location, Consumer<Boolean>>> oldListeners;
+        List<Tuple<Long, Consumer<Boolean>>> oldSeqNoListeners;
         synchronized (this) {
             oldListeners = refreshListeners;
+            oldSeqNoListeners = seqNoRefreshListeners;
             refreshListeners = null;
+            seqNoRefreshListeners = null;
             closed = true;
         }
         // Fire any listeners we might have had
         fireListeners(oldListeners);
+        fireSeqNoListeners(oldSeqNoListeners);
     }
 
     /**
      * Returns true if there are pending listeners.
      */
     public boolean refreshNeeded() {
         // A null list doesn't need a refresh. If we're closed we don't need a refresh either.
+        if (indexShard != null && indexShard.indexSettings.isSegRepEnabled() && indexShard.routingEntry().primary() == false) {
```

Comment:
`return refreshListeners != null && seqNoRefreshListeners != null && false == closed` will cause non-segrep tests to fail. Because we are coupling the condition seqNoRefreshListeners != null with results of non-segrep/primary shards.


Instead I update this to:
```
if (seqNoRefreshListeners != null) {
            return false == closed;
        }
        return refreshListeners != null && false == closed;
```